# Level 2/6: How to: omgaan met missing values

In deze level behandelen we de volgende onderwerpen:
- 02-01 Percentages;
- 02-02 Weg met die rijen;
- 02-03 Weg met die kolommen;
- 02-04 Means of means;
- 02-05 Medianen to the rescue;
- 02-06 Sales calculator.

We installeren de libraries.

In [161]:
import pandas as pd

## 02-01 Percentages
Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving
In dit level ga je aan de slag met een dataset over de verkoop van video-games. Helaas komen in een dataset vaak NaN values voor. Dit kan je analyse of je plot verstoren. Daarom is het in kaart brengen van ontbrekende waardes altijd één van de eerste stappen van de data-cleaning. Het percentage ontbrekende waardes zegt namelijk iets over de datakwaliteit.

Je gaat per kolom in je dataframe het percentage onbrekende waardes berekenen. Dit kun je als volgt berekenen:

- percentage = (number of missing values / total number of datapoints ) * 100

Loop over de kolomnamen van je dataframe. Print de kolomnaam en het bijbehorende aantal NaN values.


- Bereken de som van het aantal waardes wat NaN is per kolom;
- Deel deze som door het totale aantal gegevens om het percentage ontbrekende gegevens te berekenen.

### Leerdoelen
Een overzicht maken van de hoeveelheid ontbrekende waardes in je dataset.

### Verwachte uitkomst
Percentages NaN-values per kolom:

Rank: 0.0 %
Name: 0.0 %
Platform: 0.0 %
Year: 1.6327268345583803 %
Genre: 0.0 %
Publisher: 0.3494396915290999 %
NA_Sales: 0.0 %
EU_Sales: 16.53211230268707 %
JP_Sales: 0.0 %
Other_Sales: 0.0 %
Global_Sales: 0.0 %

De het procentteken is niet verplicht om erbij weer te geven in je Jupyter Notebook.

### Nakijkcriteria
1. Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
2. Per kolom wordt het percentage NaN-values geprint. De kolommen 'Year', 'Publisher', en 'EU_Sales' bevatten NaN-values. EU_Sales heeft het hoogste percentage: 16.53211230268707 %


### Uitwerking
We lezen csv bestand in.

In [32]:
df = pd.read_csv('video_games_sales.csv')
df

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27       NaN     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00       NaN      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

We halen aantal NaN values op en berekenen percantage.

In [39]:
df.isnull().sum() / df.shape[0] * 100.00

Rank             0.000000
Name             0.000000
Platform         0.000000
Year             1.632727
Genre            0.000000
Publisher        0.349440
NA_Sales         0.000000
EU_Sales        16.532112
JP_Sales         0.000000
Other_Sales      0.000000
Global_Sales     0.000000
dtype: float64

## 02-02 Weg met die rijen!
Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving

Een manier om van je NaN values af te komen, is door simpelweg alle datapunten te verwijderen die ergens onvolledig zijn. We gaan hier deze strategie uitproberen.

- Print het aantal datapunten.
- Verwijder de rijen waar 'EU_Sales', 'Year', of 'Publisher' NaN is.
- Hoeveel datapunten heb je nu over?

### Leerdoelen

Oplossing toepassen om met missing values om te gaan.

### Verwachte uitkomst

Oorspronkelijke aantal datapunten: 16598

Na het verwijderen van rijen: 13592

### Nakijkcriteria

- Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
- Het aantal overgebleven datapunten wordt geprint en is 13592.

### Uitwerking

We maken maken een nieuwe dataframe zonder NaN.

In [47]:
new_df = df.dropna(axis = 0, how ='any')
new_df

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
5          6                                            Tetris       GB   
...      ...                                               ...      ...   
16592  16595                                          Plushees       DS   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16597  16600                                  Spirits & Spells      GBA   

         Year       Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0      Sports    Nintendo     41.49     29.02      3.77   
1      1985.0    Platform    Nintendo     29.08      3.58      6.81   
2      2008.0      Racing    Nintendo     15.85     12.88      3.79   
3      2009.0      Sports    Nintendo     15.75     11.01      3.28   
5      1989.0      Puzzle    Nintendo     23.20      2.26      4.22   
...       ...         ...         ...       ...       ...       ...   
16592  2008.0  Simulation   Destineer      0.01      0.00      0.00   
16593  2002.0    Platform       Kemco      0.01      0.00      0.00   
16594  2003.0     Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0      Racing  Activision      0.00      0.00      0.00   
16597  2003.0    Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
5             0.58         30.26  
...            ...           ...  
16592         0.00          0.01  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16597         0.00          0.01  

[13592 rows x 11 columns]

We printen de oude, de nieuwe en verschil datapunten.

In [52]:
print("Oorspronkelijke aantal datapunten:", len(df), "\nNa het verwijderen van rijen:", 
       len(new_df), "\nVerschil datapunten: ",
       (len(df)-len(new_df)))

Oorspronkelijke aantal datapunten: 16598 
Na het verwijderen van rijen: 13592 
Verschil datapunten:  3006


## 02-03 Weg met die kolommen!
Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving
In de vorige exercise heb je het probleem van ontbrekende waardes opgelost door deze rijen te verwijderen. Een andere strategie is om in plaats van rijen, de kolommen met ontbrekende waardes te verwijderen. Kun je bedenken waarom de ene of de andere strategie beter is?

- Print het aantal kolommen in je dataframe.
- Verwijder kolommen waarin NaN-values voorkomen.
- Hoeveel kolommen heb je nu over?

### Leerdoelen
Oplossing toepassen om met NaN values om te gaan.

### Nieuwe technieken
dataframe.drop
Verwijder gespecificeerde delen van pandas dataframe.

### Verwachte uitkomst
Oorspronkelijke aantal kolommen: 11

Overgebleven kolommen: 8

### Nakijkcriteria
Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
Het aantal overgebleven kolommen wordt geprint en dit zijn er 8.

### Uitwerking
We gebruiken de oorspronkelijke dataframe (df).
We printen het aantal kolommen in de oorspronkelijke dataframe.

In [77]:
print("Oorspronkelijke aantal kolommen:", len(df.columns))

Oorspronkelijke aantal kolommen: 11


We verwijderen NaN kolommen.

In [61]:
df_col = df.dropna(axis = 1, how ='any')

We printen hoeveel kolommen er nog over zijn.

In [102]:
print("Overgebleven kolommen:", len(df_col.columns))

Overgebleven kolommen: 8


## 02-04 Means of Means
Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving
Een methode om al je datapunten te behouden, is door de ontbrekende waardes in te vullen met een aannemelijke waarde. Je zou ontbrekende waardes kunnen invullen met het gemiddelde van de kolom. Een voordeel hiervan is dat de 'mean' (het gemiddelde) van de kolom hetzelfde blijft. We gaan deze strategie toepassen op de kolom 'EU_Sales'.

- Print het gemiddelde van de kolom EU_Sales.
- Vul de NaN values in deze kolom met dit gemiddelde.
- Print het percentage NaN values. Is het gelukt?
- Print het gemiddelde van de kolommen. Is deze onveranderd gebleven?

### Leerdoelen
Oplossing voor NaN values toepassen.

### Nieuwe technieken
Met de functie fillna van pandas kun je NaN values opvullen met een gekozen waarde.

### Verwachte uitkomst
- Gemiddelde EU Sales: 0.14876714306337524
- Percentage NaN na opvullen: 0.0 %
- Nieuw gemiddelde: 0.14876714306337518

### Nakijkcriteria
- Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
- Het gemiddelde voor en na het invullen van NaN-values is 0.14876714306337518


### Uitwerking
We lezen de oorspronkelijke dataframe in (df).<br>
We printen het gemiddelde van de kolom EU_Sales.

In [152]:
df2_mean_old = pd.read_csv('video_games_sales.csv')
df2_mean_old

Rank                                              Name Platform  \
0          1                                        Wii Sports      Wii   
1          2                                 Super Mario Bros.      NES   
2          3                                    Mario Kart Wii      Wii   
3          4                                 Wii Sports Resort      Wii   
4          5                          Pokemon Red/Pokemon Blue       GB   
...      ...                                               ...      ...   
16593  16596                Woody Woodpecker in Crazy Castle 5      GBA   
16594  16597                     Men in Black II: Alien Escape       GC   
16595  16598  SCORE International Baja 1000: The Official Game      PS2   
16596  16599                                        Know How 2       DS   
16597  16600                                  Spirits & Spells      GBA   

         Year         Genre   Publisher  NA_Sales  EU_Sales  JP_Sales  \
0      2006.0        Sports    Nintendo     41.49     29.02      3.77   
1      1985.0      Platform    Nintendo     29.08      3.58      6.81   
2      2008.0        Racing    Nintendo     15.85     12.88      3.79   
3      2009.0        Sports    Nintendo     15.75     11.01      3.28   
4      1996.0  Role-Playing    Nintendo     11.27       NaN     10.22   
...       ...           ...         ...       ...       ...       ...   
16593  2002.0      Platform       Kemco      0.01      0.00      0.00   
16594  2003.0       Shooter  Infogrames      0.01      0.00      0.00   
16595  2008.0        Racing  Activision      0.00      0.00      0.00   
16596  2010.0        Puzzle    7G//AMES      0.00       NaN      0.00   
16597  2003.0      Platform     Wanadoo      0.01      0.00      0.00   

       Other_Sales  Global_Sales  
0             8.46         82.74  
1             0.77         40.24  
2             3.31         35.82  
3             2.96         33.00  
4             1.00         31.37  
...            ...           ...  
16593         0.00          0.01  
16594         0.00          0.01  
16595         0.00          0.01  
16596         0.00          0.01  
16597         0.00          0.01  

[16598 rows x 11 columns]

We printen de oude gemiddelde.

In [153]:
mean_old = df2_mean_old["EU_Sales"].mean()
print("Gemiddelde EU Sales: ", mean_old)

Gemiddelde EU Sales:  0.14876714306337524


We lezen nogmaals de csv bestand in.

In [154]:
df2_mean_new = pd.read_csv('video_games_sales.csv')

We vervangen de NaN values met de oude gemiddelde.

In [158]:
df2_mean_new["EU_Sales"].fillna(mean_old, inplace = True)

We printen de percentage NaN na opvulling.

In [159]:
print("Percentage NaN na opvullen: ", df['EU_Sales'].isna().sum() * 100.00, '%')

Percentage NaN na opvullen:  0.0 %


We printen de nieuwe gemiddelde na opvulling.

In [160]:
mean_new = df2_mean_new["EU_Sales"].mean()
print("Gemiddelde EU Sales: ", mean_new)

Gemiddelde EU Sales:  0.14876714306337518


## 02-05 Medianen To the Rescue

Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving
Een andere optie om ontbrekende waarden in te vullen, is door middel van de mediaan. Als je al je waarnemingen in je kolom op volgorde zet van klein naar groot, is de mediaan de middelste waarneming. Als je data niet gelijk verdeeld is, dan kan de mediaan een beter beeld geven rond welk getal de meeste waardes voorkomen.


- Print het gemiddelde van de kolom EU_Sales.
- Vul de NaN values in deze kolom met de mediaan van deze kolom.
- Print het percentage NaN values. Is het gelukt?
- Print het nieuwe gemiddelde van de kolommen. Is het gemiddelde sterk veranderd?

### Leerdoelen
Oplossing voor NaN values toepassen.

### Nieuwe technieken
Bereken de mediaan van een pandas dataframe.

### Herhaalde technieken
Met de functie fillna van pandas kun je NaN values opvullen met een gekozen waarde.

#### Verwachte uitkomst
- Gemiddelde EU Sales: 0.14876714306337524
- Percentage NaN na opvullen: 0.0
- Nieuw gemiddelde: 0.1274792143631763

### Nakijkcriteria
- Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
- Het gemiddelde verandert na het invullen invoegen van de mediaan van 0.14876714306337524 naar 0.1274792143631763

### Uitwerking
We lezen de csv bestand in.

In [205]:
df3 = pd.read_csv('video_games_sales.csv')

We berekenen de oude gemidddelde.

In [206]:
mean_new = df3["EU_Sales"].mean()
print("Gemiddelde EU Sales: ", mean_new)

Gemiddelde EU Sales:  0.14876714306337524


We berekenen de mediaan

In [207]:
median_old = df3["EU_Sales"].median()
print("Gemiddelde EU Sales: ", median_old)

Gemiddelde EU Sales:  0.02


We lezen nogmaals de csv bestand in.

In [208]:
df4 = pd.read_csv('video_games_sales.csv')

We vervangen de NaN door de mediaan.

In [209]:
df4["EU_Sales"].fillna(median_old, inplace = True)

We printen de percentage NaN na opvullen.

In [210]:
print("Percentage NaN na opvullen: ", df4['EU_Sales'].isna().sum() * 100.00, '%')

Percentage NaN na opvullen:  0.0 %


We berekenen de nieuwe gemiddelde uit na opvullen.

In [211]:
mean_new = df4["EU_Sales"].mean()
print("Nieuw gemiddelde: ", mean_new)

Nieuw gemiddelde:  0.1274792143631763


## 02-06 Sales Calculator
Hieronder een beschrijving en de uitwerking van de opdracht.

### Omschrijving
In de vorige exercises heb je het probleem van ontbrekende waardes opgelost door ze te verwijderen of een inschatting te maken van wat de waarde zou kunnen zijn. In sommige gevallen kan je de daadwerkelijke waarde afleiden uit de data zelf.

In deze video-game dataset blijken de global sales een optelling te zijn van alle andere sales. Doordat we dit patroon hebben ontdekt, kunnen we de ontbrekende EU_Sales berekenen.

Global_Sales = NA_Sales + EU_Sales + JP_Sales + Other_Sales<br>
EU_Sales = Global_Sales - NA_Sales - JP_Sales - Other_Sales<br>
Bereken de variabele EU_Sales zodat we geen missing values meer hebben!

In de vorige exercises heb je gekeken hoe het gemiddelde verandert als je NaN-values met een geschatte waarde opvult. Nu dat je de echte waardes heb berekend, kun je het echte gemiddelde uitrekenen. Doe dit en print deze waarde.

### Leerdoelen
Oplossing voor missing values toepassen.

### Verwachte uitkomst
Echte gemiddelde EU_Sales: 0.146......

### Nakijkcriteria
Als je alle libraries uit "requirements.txt" hebt geïnstalleerd, dan kun je de Jupyter Notebook runnen.
Het gemiddelde wordt geprint en is 0.146....

### Uitwerking
We lezen de csv bestand in.

In [337]:
df5 = pd.read_csv('video_games_sales.csv')

We berekenen de values van "EU_Sales".

In [338]:
cal_eu_sales = df5['Global_Sales'] - df5['NA_Sales'] - df5['JP_Sales'] - df5['Other_Sales']

We vervangen NaN door berekende "EU_Sales".

In [339]:
df5['EU_Sales'] = df5['EU_Sales'].fillna(value=cal_eu_sales, inplace=False)

We bereken de echte mean van "EU_Sales".

In [340]:
real_mean = df5["EU_Sales"].mean()
print("Echte gemiddelde EU_Sales: ", real_mean)

Echte gemiddelde EU_Sales:  0.1466622484636703
